File: Maximum Difference Between Even and Odd Frequency II/Maximum Difference Between Even and Odd Frequency II.py
# Solution for Maximum Difference Between Even and Odd Frequency II in PY

class Solution:
    def getState(self, countA: int, countB: int) -> int:
        parityA = countA % 2
        parityB = countB % 2
        if parityA == 0 and parityB == 0:
            return 0
        if parityA == 0 and parityB == 1:
            return 1
        if parityA == 1 and parityB == 0:
            return 2
        return 3
        # return (parityA << 1) | parityB

    def maxDifference(self, s: str, k: int) -> int:
        n = len(s)
        maxDiff = float('-inf')

        for a in '01234':
            for b in '01234':
                if a == b:
                    continue

                minPrefixDiffForState = [float('inf')] * 4

                totalCountA = totalCountB = 0
                leftCountA = leftCountB = 0
                left = -1
                right = 0

                while right < n:
                    if s[right] == a:
                        totalCountA += 1
                    if s[right] == b:
                        totalCountB += 1

                    while right - left >= k and (totalCountB - leftCountB) >= 2 and (totalCountA - leftCountA) >= 1:
                        leftState = self.getState(leftCountA, leftCountB)
                        minPrefixDiffForState[leftState] = min(
                            minPrefixDiffForState[leftState],
                            leftCountA - leftCountB
                        )
                        left += 1
                        if left < n:
                            if s[left] == a:
                                leftCountA += 1
                            if s[left] == b:
                                leftCountB += 1

                    currentState = self.getState(totalCountA, totalCountB)
                    requiredPrevState = currentState ^ 2

                    if minPrefixDiffForState[requiredPrevState] != float('inf'):
                        diff = (totalCountA - totalCountB) - minPrefixDiffForState[requiredPrevState]
                        maxDiff = max(maxDiff, diff)

                    right += 1

        return maxDiff

File: Maximum Difference Between Even and Odd Frequency II/Maximum Difference Between Even and Odd Frequency II.cpp
// Solution for Maximum Difference Between Even and Odd Frequency II in CPP

class Solution {
public:

    // Determine the parity state of counts: 
    // 0 = even-even, 1 = even-odd, 2 = odd-even, 3 = odd-odd
    int getState(int countA, int countB) {
        int parityA = countA % 2;
        int parityB = countB % 2;
        if (parityA == 0 && parityB == 0) return 0; // even even (00)
        if (parityA == 0 && parityB == 1) return 1; // even odd (01)
        if (parityA == 1 && parityB == 0) return 2; // odd even (10)
        return 3; // odd odd (11)
        // return (parityA << 1) | parityB; // combine parity bits
    }

    int maxDifference(string s, int k) {
        int n = s.length();
        int maxDiff = INT_MIN;

        // Try all pairs of characters a and b ('0' to '4'), where a != b
        for (char a = '0'; a <= '4'; a++) {
            for (char b = '0'; b <= '4'; b++) {
                if (a == b) continue;

                // Stores the minimum value of (countA - countB) for each state
                vector<int> minPrefixDiffForState(4, INT_MAX);

                // Total counts of a and b up to the current index
                int totalCountA = 0, totalCountB = 0;

                // Counts of a and b from left index
                int leftCountA = 0, leftCountB = 0;

                int left = -1, right = 0;

                while (right < n) {
                    // Update counts with current character
                    totalCountA += (s[right] == a);
                    totalCountB += (s[right] == b);

                    // Shrink window from left while maintaining minimum required length
                    while (right - left >= k && totalCountB - leftCountB >= 2 && totalCountA - leftCountA >= 1) {
                        int leftState = getState(leftCountA, leftCountB);
                        minPrefixDiffForState[leftState] = min(minPrefixDiffForState[leftState], leftCountA - leftCountB);

                        // Move left pointer and update counts
                        left++;
                        leftCountA += (s[left] == a);
                        leftCountB += (s[left] == b);
                    }

                    // Current window state
                    int currentState = getState(totalCountA, totalCountB);

                    // Find corresponding previous state with a=odd, b=even (using XOR with 2 flips parity of a)
                    int requiredPrevState = currentState ^ 2;

                    if (minPrefixDiffForState[requiredPrevState] != INT_MAX) {
                        int diff = (totalCountA - totalCountB) - minPrefixDiffForState[requiredPrevState];
                        maxDiff = max(maxDiff, diff);
                    }

                    right++;
                }
            }
        }

        return maxDiff;
    }
};
